# Facility Location


## Objective and Prerequisites

Facility location problems can be commonly found in many industries, including logistics and telecommunications. In this example, we’ll show you how to tackle a facility location problem that involves determining the number and location of warehouses that are needed to supply a group of supermarkets.


## Motivation

The study of facility location problems - also known as "location analysis" [1] - is a branch of operations research and computational geometry concerned with the optimal placement of facilities to minimize transportation costs while considering factors like safety (e.g. by avoiding placing hazardous materials near housing) and the location of competitors' facilities.

The Fermat-Weber problem, formulated in the 17'th century, was one of the first facility location problems ever devised.
The Fermat-Weber problem can be described as follows: Given three points in a plane, find a fourth point such that the sum of its distances to the three given points is minimal. This problem can be viewed as a variation of the facility location problem, where the assumption is made that the transportation costs per distance are the same for all destinations.

Facility location problems have applications in a wide variety of industries. For supply chain management and logistics, this problem can be used to find the optimal location for stores, factories, warehouses, etc. Other applications range from public policy (e.g. positioning police officers in a city), telecommunications (e.g. cell towers in a network), and even particle physics (e.g. separation distance between repulsive charges). Another application of the facility location problem is to determine the locations for natural gas transmission equipment. Finally, facility location problems can be applied to cluster analysis.


## Problem Description

A large supermarket chain in the UK needs to build warehouses for a set of supermarkets it is opening in Northern England. The locations of the supermarkets have been identified, but the locations of the warehouses have yet to be determined.

Several good candidate locations for the warehouses have been identified, but decisions must be made regarding
how many warehouses to open and at which candidate locations to build them.

Opening many warehouses would be advantageous as this would reduce the average distance a truck has to drive from the warehouse to the supermarket, and hence reduce the delivery cost. However, opening a warehouse has a fixed cost associated with it.

In this example, our goal is to find the optimal tradeoff between delivery costs and the costs of building new facilities.


## Solution Approach

Mathematical programming is a declarative approach where the modeler formulates a mathematical optimization model that captures the key aspects of a complex business problem.

A mathematical optimization model has five components, namely:

- Sets and indices.
- Parameters.
- Decision variables.
- Objective function(s).
- Constraints.

We now present a MIP formulation for the facility location problem.


## Model Formulation

### Sets and Indices

$i \in I$: Index and set of supermarket (or customer) locations.

$j \in J$: Index and set of candidate warehouse (or facility) locations.

### Parameters

$f_{j} \in \mathbb{R}^+$: Fixed cost associated with constructing facility $j \in J$.

$d_{i,j} \in \mathbb{R}^+$: Distance between facility $j \in J$ and customer $i \in I$.

$c_{i,j} \in \mathbb{R}^+$: Cost of shipping between candidate facility site $j \in J$ and customer location $i \in I$. We assume that this cost is proportional to the distance between the facility and the customer. That is, $c_{i,j} = \alpha \cdot d_{i,j}$, where $\alpha$ is the cost per mile of driving, adjusted to incorporate the average number of trips a delivery truck would be expected to make over a five year period.

### Decision Variables

$select_{j} \in \{0, 1 \}$: This variable is equal to 1 if we build a facility at candidate location $j \in J$; and 0 otherwise.

$0 \leq assign_{i,j} \leq 1$: This non-negative continuous variable determines the fraction of supply received by customer $i \in I$ from facility $j \in J$.

### Objective Function

- **Total costs**. We want to minimize the total cost to open and operate the facilities. This is the sum of the cost of opening facilities and the cost related to shipping between facilities and customers. This total cost measures the tradeoff between the cost of building a new facility and the total delivery cost over a five year period.

\begin{equation}
\text{Min} \quad Z = \sum_{j \in J} f_{j} \cdot select_{j} + \sum_{j \in J} \sum_{i \in I} c_{i,j} \cdot assign_{i,j}
\tag{0}
\end{equation}

### Constraints

- **Demand**. For each customer $i \in I$ ensure that its demand is fulfilled. That is, the sum of the fraction received from each facility for each customer must be equal to 1:

\begin{equation}
\sum_{j \in J} assign_{i,j} = 1 \quad \forall i \in I
\tag{1}
\end{equation}

- **Shipping**. We need to ensure that we only ship from facility $j \in J$, if that facility has actually been built.

\begin{equation}
assign_{i,j} \leq select_{j} \quad \forall i \in I \quad \forall j \in J
\tag{2}
\end{equation}


## Python Implementation

This example considers two supermarkets and nine warehouse candidates. The coordinates of each supermarket are provided in the following table.

| <i></i>       | Coordinates |
| ------------- | ----------- |
| Supermarket 1 | (0,1.5)     |
| Supermarket 2 | (2.5,1.2)   |

The following table shows the coordinates of the candidate warehouse sites and the fixed cost of building the warehouse in millions of GBP.

| <i></i>     | coordinates | fixed cost |
| ----------- | ----------- | ---------- |
| Warehouse 1 | (0,0)       | 3          |
| Warehouse 2 | (0,1)       | 2          |
| Warehouse 3 | (0,2)       | 3          |
| Warehouse 4 | (1,0)       | 1          |
| Warehouse 5 | (1,1)       | 3          |
| Warehouse 6 | (1,2)       | 3          |
| Warehouse 7 | (2,0)       | 4          |
| Warehouse 8 | (2,1)       | 3          |
| Warehouse 9 | (2,2)       | 2          |

The cost per mile is one million GBP.

## Python Implementation

We now import the Gurobi Python Module and other Python libraries. Then, we initialize the data structures with the given data.


In [2]:
from itertools import product
from math import sqrt
import pyomo.environ as pyo

# Parameters
customers = [(0, 1.5), (2.5, 1.2)]
facilities = [(0, 0), (0, 1), (0, 2), (1, 0), (1, 1), (1, 2), (2, 0), (2, 1), (2, 2)]
setup_cost = [3, 2, 3, 1, 3, 3, 4, 3, 2]
cost_per_mile = 1

### Preprocessing

We define a function that determines the Euclidean distance between each facility and customer sites. In addition, we compute key parameters required by the MIP model formulation of the facility location problem.


In [20]:
# This function determines the Euclidean distance between a facility and customer sites.


def compute_distance(loc1, loc2):
    dx = loc1[0] - loc2[0]
    dy = loc1[1] - loc2[1]
    return sqrt(dx * dx + dy * dy)


# Compute key parameters of MIP model formulation

num_facilities = len(facilities)
num_customers = len(customers)
cartesian_prod = list(product(range(num_customers), range(num_facilities)))

# Compute shipping costs

shipping_cost = {
    (c, f): cost_per_mile * compute_distance(customers[c], facilities[f])
    for c, f in cartesian_prod
}
# shipping_cost = {}
# for c, f in cartesian_prod:
#     if c not in shipping_cost:
#         shipping_cost[c] = {}
#     shipping_cost[c][f] = cost_per_mile * compute_distance(customers[c], facilities[f])
# # = {}
shipping_cost[0, 0]

1.5

### Model Deployment

We now determine the MIP model for the facility location problem, by defining the decision variables, constraints, and objective function. Next, we start the optimization process and Gurobi finds the plan to build facilities that minimizes total costs.


In [26]:
# MIP  model formulation

m = pyo.ConcreteModel("facility_location")


m.select = pyo.Var(list(range(num_facilities)), domain=pyo.Binary, name="Select")

m.assign = pyo.Var(
    list(range(num_customers)),
    list(range(num_facilities)),
    bounds=(None, 1),
    domain=pyo.NonNegativeReals,
    name="Assign",
)

m.setup_to_ship_constraints = pyo.ConstraintList()
for c, f in cartesian_prod:
    m.setup_to_ship_constraints.add(m.assign[c, f] <= m.select[f])


m.demand_constraints = pyo.ConstraintList()
for c in range(num_customers):
    m.demand_constraints.add(sum(m.assign[c, f] for f in range(num_facilities)) == 1)


def OBJ(m):
    total_shipping_cost = pyo.sum_product(m.assign, shipping_cost, index=cartesian_prod)
    total_setup_cost = pyo.sum_product(
        m.select, setup_cost, index=range(num_facilities)
    )
    return total_setup_cost + total_shipping_cost


m.OBJ = pyo.Objective(rule=OBJ, sense=pyo.minimize)

In [27]:
# solve
pyo.SolverFactory("cbc").solve(m).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 4.72371291
  Upper bound: 4.72371291
  Number of objectives: 1
  Number of constraints: 20
  Number of variables: 27
  Number of binary variables: 9
  Number of integer variables: 9
  Number of nonzeros: 27
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  User time: -1.0
  System time: 0.0
  Wallclock time: 0.0
  Termination condition: optimal
  Termination message: Model was solved to optimality (subject to tolerances), and an optimal solution is available.
  Statistics: 
    Branch and

## Analysis

The result of the optimization model shows that the minimum total cost value is 4.72 million GBP. Let's see the solution that achieves that optimal result.

### Warehouse Build Plan

This plan determines at which site locations to build a warehouse.


In [28]:
# display optimal values of decision variables

for facility in m.select.keys():
    if abs(m.select[facility].value) > 1e-6:
        print(f"\n Build a warehouse at location {facility + 1}.")


 Build a warehouse at location 4.


### Shipment Plan

This plan determines the percentage of shipments to be sent from each facility built to each customer.


In [29]:
# Shipments from facilities to customers.

for customer, facility in m.assign.keys():
    if abs(m.assign[customer, facility].value) > 1e-6:
        print(
            f"\n Supermarket {customer + 1} receives {round(100*m.assign[customer, facility].value, 2)} % of its demand  from Warehouse {facility + 1} ."
        )


 Supermarket 1 receives 100.0 % of its demand  from Warehouse 4 .

 Supermarket 2 receives 100.0 % of its demand  from Warehouse 4 .


## Conclusion

In this example, we addressed a facility location problem where we want to build warehouses to supply a large number of supermarkets while minimizing the fixed total costs of building warehouses and the total variable shipping costs from warehouses to supermarkets. We learned how to formulate the problem as a MIP model.


## References

[1] Laporte, Gilbert, Stefan Nickel, and Saldanha da Gama, Francisco. Location Science. Springer, 2015.
